feat: allow python expressions in --set-resources

johanneskoester opened the pull request:
### Description

<!--Add a description of your PR here-->

### QC
<!-- Make sure that you can tick the boxes below. -->

* [ ] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [ ] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

dlaehnemann commented:
I now added and cleaned up some of the documentation. Also, I amended the profile test case to cover the different profile resource keywords and actually test them. This exposed two things for `--set-threads`:

1. Using a callable as suggested in the docs for `threads:`, does not work for `set-threads:` in a profile. I am using `input.size_mb`, as the docs suggest that only `wildcards, [input]` are valid usable arguments for the callable here.
2. When only specifying a number for a `rule x` via `set-threads:`, this does get parsed correctly. But interestingly, it only gets used, if the `rule x` already has a value specified with the `threads:` directive. Otherwise, whatever you specify via `set-threads:` in the profile is just ignored.

I think these two things make sense to already fix here, right?

johanneskoester commented:
> I now added and cleaned up some of the documentation. Also, I amended the profile test case to cover the different profile resource keywords and actually test them. This exposed two things for `--set-threads`:
> 
>     1. Using a callable as suggested in the docs for `threads:`, does not work for `set-threads:` in a profile. I am using `input.size_mb`, as the docs suggest that only `wildcards, [input]` are valid usable arguments for the callable here.
> 
>     2. When only specifying a number for a `rule x` via `set-threads:`, this does get parsed correctly. But interestingly, it only gets used, if the `rule x` already has a value specified with the `threads:` directive. Otherwise, whatever you specify via `set-threads:` in the profile is just ignored.
> 
> 
> I think these two things make sense to already fix here, right?

Yes, fixing it already makes sense.
